### PR TITLE
Add Rails 7 & Ruby 3.1 support

### LINKED ISSUE #352: Add Support for Rails 7
Add support for Rails 7.0 and Ruby 3.1.

Note: It appears that v0.16.0 will work fine with Rails 7 so long as you are using the appropriate version of heartcombo/devise. This PR just adds Rails 7 and Ruby 3.1 to the build matrix so we can run tests for those versions.

- [x] Update ruby version
- [x] Update dependencies
- [x] Add Rails 7 to build matrix
- [x] Add Ruby 3.1 to build matrix

### PR DESCRIPTION
**Pull request by olbrich**
Fixes #352

- [ ] Depends on https://github.com/blowmage/minitest-rails/pull/232 & https://github.com/blowmage/minitest-rails/issues/234
- [ ] Depends on https://jira.mongodb.org/browse/RUBY-2834
- [x] Add ruby 3.1 to build matrix

It looks like MongoDB hasn't even started work on supporting Rails 7, so we may have to drop mongoid support for Rails 7.0 for now.

Note that Rails 7.0 and Ruby 3.1 are incompatible, we will be expecting that combination to fail until  Rails 7.0.1 comes out. See https://github.com/rails/rails/issues/43998 and https://github.com/rails/rails/pull/43951

### CODE REVIEW

**Comment by dillonwelch** [comment]
Everything looks fine to me.

We shouldn't have to officially drop anything on our end because the `mongoid` gem has a `< 6.2` requirement for `activemodel` on their end, which would already prevent anyone from using it with Rails 7.0.

https://github.com/mongodb/mongoid/blob/master/mongoid.gemspec#L36

**Comment by dillonwelch** [comment]
https://twitter.com/rails/status/1479214221506322433 patch out that should help with this

File: Appraisals [hunk 1]
```diff
@@ -9,18 +9,41 @@ appraise 'rails-6.0' do
...
+  # but is used by the `mail` gem. So we need to add them as dependencies until
+  # `mail` is fixed: https://github.com/mikel/mail/pull/1439 Taken from
+  # https://github.com/rails/rails/pull/42366
+  gem 'net-smtp', require: false
+end
+
+appraise 'rails-7.0' do
+  gem 'railties', '~> 7.0.0'
+  gem 'devise', '>= 4.8.1' # devise introduced Rails 7 support
+  gem 'minitest-rails', github: 'blowmage/minitest-rails', branch: 'master' # temporary until official release of Rails 7 support for minitest-rails
```
Comment (djpremier) [inline Appraisals:33]: The official release already available:

https://github.com/minitest/minitest-rails/releases/tag/v7.0.0

**Comment by djpremier** [comment]
> @dillonwelch Agreed, we just won't be able to support the specific combination of Rails 7 and Mongoid until they provide Rails 7 support. I'm OK with shipping this when the minitest support for Rails 7 ships.

Mongo also already support Rails 7.0

https://github.com/mongodb/mongoid/pull/5122

**Comment by evolve2k** [comment]
Checking in on this one. Would be good to have this merged.

**Comment by chipach** [comment]
@djpremier @olbrich - Any thoughts on getting this merged in?  Looks like it is failing a test, but the logs are gone from Github.  This would save us a bunch of work to fork/reimplement chunks of it.  Thank you!

**Comment by djpremier** [comment]
> @djpremier @olbrich - Any thoughts on getting this merged in? Looks like it is failing a test, but the logs are gone from Github. This would save us a bunch of work to fork/reimplement chunks of it. Thank you!

Hi @chipach, I will be recreating this branch with the necessary adjustments since the original author has not interacted anymore.

**Pull request closed without merging**